MediaRecorder.bs

```diff
@@ -196,23 +216,44 @@ interface MediaRecorder : EventTarget {
     <li>If {{state}} is not {{inactive}}, throw an {{InvalidStateError}}
     {{DOMException}} and abort these steps.</li>
 
-    <li>If the {{MediaRecorder/stream}}'s <a>isolation properties</a> disallow access
+    <li>If the {{stream}}'s <a>isolation properties</a> disallow access
     from this {{MediaRecorder}}, throw a {{SecurityError}} {{DOMException}} and
     abort these steps.</li>
 
+    <li>If the {{stream}} is {{inactive}}, throw a {{NotSupportedError}}
+    {{DOMException}} and abort these steps.</li>
+
+    <li>Let |tracks| be the set of {{live}} tracks in the {{stream}}'s
+    [=track set=].</li>
+
+    <li>If the [=constrainedMimeType=] internal property specifies a media type,
+    container, or codec such that the User Agent cannot record a track in
+    |tracks| with it, throw a {{NotSupportedError}} {{DOMException}} and abort
+    these steps.</li>
+
+    <li>Set {{mimeType}} to the [=constrainedMimeType=] internal property,
+    extended by the media type, subtype and codecs parameter reflecting the
+    configuration used by the MediaRecorder to record all tracks in |tracks|.
+    The media type, subtype and any parameters in the [=constrainedMimeType=]
+    internal property, MUST NOT be changed. {{mimeType}} MAY be extended with
+    the profiles parameter [[RFC6381]] or further codec-specific parameters to
+    avoid ambiguity.</li>
+
     <li>Set {{state}} to {{recording}}, and run the following steps in parallel:
     <ol>
-      <li>Start gathering the data into a {{Blob}} <var>blob</var> and queue a
-      task, using the DOM manipulation task source, to <a>fire an event</a>
-      named <a>start</a> at <var>target</var>, then return <code>undefined</code>.
-      </li>
-
-      <li>If at any point the {{MediaRecorder/stream}}'s <a>isolation properties</a>
-      change so that {{MediaRecorder}} is no longer allowed access to it, the
-      UA MUST immediately stop gathering data, discard any data that it has
-      gathered, and queue a task, using the DOM manipulation task source, that
-      runs the following steps:
+      <li>Start recording all tracks in |tracks| and gather the data into a
+      {{Blob}} <var>blob</var> and queue a task, using the DOM manipulation task
+      source, to <a>fire an event</a> named <a>start</a> at <var>target</var>,
+      then return <code>undefined</code>.</li>
+
+      <li>If at any point the {{stream}}'s <a>isolation properties</a> change so
+      that {{MediaRecorder}} is no longer allowed access to it, the UA MUST
+      immediately stop gathering data, discard any data that it has gathered,
+      and queue a task, using the DOM manipulation task source, that runs the
+      following steps:
       <ol>
+        <li>Set {{mimeType}} to the value of the [=constrainedMimeType=]
+        internal property.</li>
```

Comment:
Do browsers currently do this?

Reply:
Firefox does sort of, it sets `mimeType` to "" before firing "stop", https://searchfox.org/mozilla-central/rev/0b7007a23bc16c857f894140e12f307bfeef2fdd/dom/media/MediaRecorder.cpp#396-397

Chrome doesn't. It selects `mimeType` in the constructor and doesn't change it even if using a different configuration to encode tracks.

Here's a modified fiddle showing this behavior, https://jsfiddle.net/pehrsons/1Lznhjyg/

Exemplified, Firefox:
```
[MediaRecorder "" [audio,video]]: ctor-mime: ""; recording-mime: "video/webm"; stopped-mime: ""; "blob-mime: "video/webm"; recorded tracks: [audio,video]
[MediaRecorder "audio/webm;codecs=opus" [video]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "video/webm"; stopped-mime: ""; "blob-mime: "video/webm"; recorded tracks: [video]
[MediaRecorder "video/webm;codecs=opus,vp8" [audio,video]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm"; stopped-mime: ""; "blob-mime: "video/webm"; recorded tracks: [audio,video]
```

Chrome:
```
[MediaRecorder "" [audio,video]]: ctor-mime: ""; recording-mime: ""; stopped-mime: ""; "blob-mime: ""; recorded tracks: [audio,video]
[MediaRecorder "audio/webm;codecs=opus" [video]]: ctor-mime: "audio/webm;codecs=opus"; recording-mime: "audio/webm;codecs=opus"; stopped-mime: "audio/webm;codecs=opus"; "blob-mime: "audio/webm;codecs=opus"; recorded tracks: [video]
[MediaRecorder "video/webm;codecs=opus,vp8" [audio,video]]: ctor-mime: "video/webm;codecs=opus,vp8"; recording-mime: "video/webm;codecs=opus,vp8"; stopped-mime: "video/webm;codecs=opus,vp8"; "blob-mime: "video/webm;codecs=opus,vp8"; recorded tracks: [audio,video]
```

Reply:
If the idea is that the codec doesn't change, then I wouldn't expect it to change on stop either. So I'd go with Chrome here.

Reply:
Resolved offline. Since `stop()` is not terminal, it makes some sense to revert to `constrainedMimeType` here to avoid confusion about whether the returned value is a constrained mimeType or merely a current mimeType.